File: Ex-26/c++/Pessoa.cpp
#include "Pessoa.h"
#include "Familia.h"

#include <iostream>
#include <string>
using namespace std;

Pessoa::Pessoa(const char* n, Pessoa* pMae, Pessoa* pPai) :
nome(n),
mae(pMae),
pai(pPai) {
    pFamilia = NULL;
}

Pessoa::Pessoa(char* n, Pessoa* pMae, Pessoa* pPai) :
nome(n),
mae(pMae),
pai(pPai) {
    pFamilia = NULL;
}

Pessoa::~Pessoa() {
    pai = NULL;
    mae = NULL;
    pFamilia = NULL;
}

string Pessoa::getNome() {
    return nome;
}

string Pessoa::getNomeCompleto() {
    string nomeCompleto = nome;

    if (pFamilia != NULL) {
        nomeCompleto += " ";
        nomeCompleto += pFamilia->getSobrenome();
    }

    return nomeCompleto;
}

void Pessoa::setNome(string* nome) {
    this->nome = *nome;
}

void Pessoa::imprimeNomeCompleto() const {
    if (pFamilia)
        cout << nome << " " << pFamilia->getSobrenome() << endl;
    else
        cout << nome << endl;
}

void Pessoa::setFamilia(Familia* pFm) {
    this->pFamilia = pFm;
}

void Pessoa::setMae(Pessoa* pMae) {
    if (pMae == NULL) {
        cout << "Ponteiro para mae NULL" << endl;
        return;
    }
    this->mae = pMae;
}

void Pessoa::setPai(Pessoa* pPai) {
    if (pPai == NULL) {
        cout << "Ponteiro para pai NULL" << endl;
        return;
    }
    this->pai = pPai;
}

void Pessoa::adicionarFilho(Pessoa* pFilho) {
    if (pFilho == NULL) {
        cout << "Ponteiro para filho NULL" << endl;
        return;
    }
    listaFilhos.push_back(pFilho);
}

void Pessoa::listarFilhos() {
    if (listaFilhos.size() < 1) {
        cout << getNomeCompleto() << " não possui filhos!" << endl;
        return;
    }

    cout << "Filhos de " << getNomeCompleto() << ":" << endl;

    for (int i = 0; i < listaFilhos.size(); i++) {
        cout << " - " << listaFilhos[i]->getNomeCompleto() << endl;
    }
}

File: Ex-26/ex_26_java/Pessoa.java
/**
 * Representa um indivíduo.
 * Uma Pessoa pode existir independentemente de uma Família.
 */
public class Pessoa {
    private String nome;
    private Familia familia; // Referência à família da qual a pessoa faz parte
    private Pessoa pai;
    private Pessoa mae;

    /**
     * Construtor para criar uma Pessoa apenas com o nome.
     * As associações com família e pais são feitas posteriormente.
     * @param nome O primeiro nome da pessoa.
     */
    public Pessoa(String nome) {
        this.nome = nome;
        this.familia = null; // Inicialmente, a pessoa não pertence a nenhuma família
        this.pai = null;
        this.mae = null;
    }

    public String getNome() {
        return nome;
    }

    public void setPai(Pessoa pai) {
        this.pai = pai;
    }
    
    public void setMae(Pessoa mae) {
        this.mae = mae;
    }
    
    // Método package-private para que apenas a classe Familia possa atribuir a família.
    void setFamilia(Familia familia) {
        this.familia = familia;
    }

    /**
     * Retorna o nome completo da pessoa, combinando seu nome com o sobrenome da família.
     * @return O nome completo (ex: "Bart Simpson") ou apenas o primeiro nome se não estiver em uma família.
     */
    public String getNomeCompleto() {
        if (this.familia != null && this.familia.getSobrenome() != null) {
            return this.nome + " " + this.familia.getSobrenome();
        }
        return this.nome; // Retorna apenas o nome se não houver família associada
    }

    @Override
    public String toString() {
        // Facilita a impressão do objeto Pessoa
        return getNomeCompleto();
    }
}